### PR TITLE
master <-- JCSP-381:DC,Cloud: Add an attachment

### PR DESCRIPTION
**Pull request by NataliiaShavshyna**
JCSP-381:DC,Cloud: Add an attachment

### CODE REVIEW

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: `val src` is unused

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: We don't have to care if it's `span`
```suggestion
            "//*[contains(text(),'Attachments')])")))
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]:
```suggestion
            attachScreenShot(loadedIssuePage)
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: You can use the Groovy-style call:
```suggestion
        meter.measure(ActionType("Attach ScreenShot") { }) {
            attachScreenShot.pasteScreenShot()
            attachScreenShot.waitForTheScreenShotAttached()
        }
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: This doesn't return the type expected by the interface.
I highly recommend compiling and running the test locally.

**Review by dagguh — CHANGES_REQUESTED** [review]

**Comment by dagguh** [comment]
It's also blocked by CLA check. You can [set up your Git email](https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address) to match the Atlassian one. You can amend commits to update their authors.

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: removed

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: changed

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: changed

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: changed

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssuePage.kt]: removed

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: I see multiple PRs changing the same places, which generates conflicts. One trick you can use is to add these params in random order. Same goes for other "common" files like `MeasureType` or `AbstractIssuePage`

In most cases the order of these lines doesn't matter.
And if it does matter, then conflicts are legit.

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: I'd extract a method for that. Like this: https://github.com/atlassian/jces-1209/blob/fd1fd2d4da880bf2dfb5328338377b0efc1dd5ba/custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt#L34-L39

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssuePage.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssuePage.kt]: This is a copy paste of `DcAddScreenShot`

**Review by dagguh — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt:10]: In general there's no reason to make a class `abstract` if it doesn't have any `abstract` members. If you want to open it for inheritance (which should be replaced by composition anyway), then you can use `open class`.

**Review by dagguh — DISMISSED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt:68]: 👍

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: variable is not used
```suggestion
        (driver as TakesScreenshot).getScreenshotAs(OutputType.BYTES)
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]:
```suggestion
        Actions(driver).keyDown(Keys.CONTROL).sendKeys("v").perform()
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: The issue could already have some attachments
It's necessary to verify the number of attachments before and after

also, the attachment could be created but not loaded - it's necessary to wait for `data-test-status="complete"` for the element with `data-testid="media-file-card-view"`
this is for Cloud

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: DC has different behavior - it's necessary to submit `upload` after the pasting
and has other locators to `wait for attachment`

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (dagguh) [reply custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: Still looks like `action = {` on my end

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (dagguh) [reply custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: It seems this was addressed, but in the parent class, so for both Cloud and DC, while the feedback was for Cloud only.
It's fine to abandon inheritance entirely and just have independent implementations for Cloud and DC. The only common part seems to be `makeScreenShot`, which could be removed from the interface entirely.

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: `sc-lbihag` looks like a very specific locator

Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: It's a `String`. A locator would have the `By` type.

**Review by dagguh — CHANGES_REQUESTED** [review]

**Review by dagguh — DISMISSED** [review]
Looks good to me. I'll request @ysemenkov's review, because he's more experienced in the DC/Cloud locators differences.
EDIT: Ah, my approval didn't discard Yevhen's requests, good 🙂

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: 1. It looks redundant in the parent class because it is not being used in the parent but adds some restrictions for children
2. screenShotXPath is being used with By.xpath each time - you can just make this one
```suggestion
    protected var countBefore: Int = 0
    abstract val screenShotLocator: By
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/WorkAnIssue.kt]: formatting - to avoid unnecessary conflicts and deltas
```suggestion
        meter.measure(ATTACH_SCREENSHOT) {
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]:
```suggestion
    protected fun waitForAttributeValueForElements(by: By, attributeName: String, attributeValue: String) {
        driver
        .findElements(by)
        .forEach {
            driver.wait(ExpectedConditions.attributeContains(it, attributeName, attributeValue))
        }
    }
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudAddScreenShot.kt]: Let's avoid implicit behavior - `countBefore` variable

Suggestion:
1. Merge methods pasteScreenShot() + waitForTheScreenShotAttached() = pasteScreenShot()
2. Make an additional argument - `waitForTheScreenShotAttached(screenShotsCountBefore: Int)`
and return this count in `pasteScreenShot()` or call `getIssueScreenShotsCount()` before `pasteScreenShot()`

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt:20]: also, it's necessary to `wait for the invisibility of modal window `

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: I can't see any elements with such locator on DC
but you could try to use this one or investigate it deeper with breakpoints in the browser
```suggestion
        waitForAttributeValueForElements(
        By.className("js-file-attachment"),
            "resolved",
            "true")
```

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]:
```suggestion
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]:
```suggestion
    protected fun getScreenShotsCount(): Int {
        return driver.findElements(screenShotLocator).size
    }
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: it's not necessary to return Int after merging methods
```suggestion
    abstract fun pasteScreenShot()
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: It's redundant because it's being checked in the previous wait

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: ExpectedConditions.and

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudAddScreenShot.kt
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudAddScreenShot.kt]: ExpectedConditions.and

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: is it possible that element can be in DOM but still invisible?

File: custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt
Comment (ysemenkov) [reply custom-vu/src/main/kotlin/jces1209/vu/page/AttachScreenShot.kt]: still need to remove
countBefore is method local variable

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (ysemenkov) [reply custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: good point. let's leave it

**Review by ysemenkov — DISMISSED** [review]
@NataliiaShavshyna don't forget to squash commits

**Review by ysemenkov — APPROVED** [review]

**Review by ysemenkov — APPROVED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt
Comment (dagguh) [reply custom-vu/src/main/kotlin/jces1209/vu/page/DcAddScreenShot.kt]: [In general, yes](https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/support/ui/ExpectedConditions.html#visibilityOf-org.openqa.selenium.WebElement-):
> Visibility means that the element is not only displayed but also has a height and width that is greater than 0.

In the case of this specific element: I don't know.

**Pull request merged**
